File: platform/lib-java/src/test/java/com/canary/platform/lib/observability/LaneStateProbeTest.java
package com.canary.platform.lib.observability;

import io.fabric8.kubernetes.api.model.EndpointAddressBuilder;
import io.fabric8.kubernetes.api.model.EndpointSubsetBuilder;
import io.fabric8.kubernetes.api.model.Endpoints;
import io.fabric8.kubernetes.api.model.EndpointsBuilder;
import io.fabric8.kubernetes.client.KubernetesClient;
import io.fabric8.kubernetes.client.dsl.MixedOperation;
import io.fabric8.kubernetes.client.dsl.NonNamespaceOperation;
import io.fabric8.kubernetes.client.dsl.Resource;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class LaneStateProbeTest {

    private SimpleMeterRegistry registry;
    private KubernetesClient client;

    @BeforeEach
    @SuppressWarnings({"unchecked", "rawtypes"})
    void setUp() {
        registry = new SimpleMeterRegistry();
        client = mock(KubernetesClient.class);
        // Stub the endpoints DSL chain so probe.start() does not NPE
        MixedOperation endpointsDsl = mock(MixedOperation.class);
        NonNamespaceOperation namespaceOp = mock(NonNamespaceOperation.class);
        Resource resource = mock(Resource.class);
        when(client.endpoints()).thenReturn(endpointsDsl);
        when(endpointsDsl.inNamespace(any())).thenReturn(namespaceOp);
        when(namespaceOp.withName(any())).thenReturn(resource);
        when(resource.get()).thenReturn(null);   // initial state: no endpoints
        when(resource.watch(any())).thenReturn(mock(io.fabric8.kubernetes.client.Watch.class));
    }

    @Test
    void startRegistersGaugesForBothLanes() {
        LaneStateProbe probe = new LaneStateProbe(client, registry, "services", "payment");
        probe.start();

        Gauge stable = registry.find("canary_lane_active")
                .tags("substrate", "http", "service", "payment", "lane", "stable").gauge();
        Gauge canary = registry.find("canary_lane_active")
                .tags("substrate", "http", "service", "payment", "lane", "canary").gauge();
        assertThat(stable).isNotNull();
        assertThat(canary).isNotNull();
        assertThat(stable.value()).isEqualTo(0.0);
        assertThat(canary.value()).isEqualTo(0.0);

        probe.close();
    }

    @Test
    void setLaneActiveTogglesGaugeValue() {
        LaneStateProbe probe = new LaneStateProbe(client, registry, "services", "payment");
        probe.start();

        probe.setLaneActive("canary", true);

        Gauge canary = registry.find("canary_lane_active")
                .tags("substrate", "http", "service", "payment", "lane", "canary").gauge();
        assertThat(canary.value()).isEqualTo(1.0);

        probe.setLaneActive("canary", false);
        assertThat(canary.value()).isEqualTo(0.0);

        probe.close();
    }

    @Test
    void hasAddressesReturnsTrueForPopulatedEndpoints() {
        Endpoints e = new EndpointsBuilder()
                .withNewMetadata().withName("svc").endMetadata()
                .addToSubsets(new EndpointSubsetBuilder()
                        .addToAddresses(new EndpointAddressBuilder().withIp("10.0.0.1").build())
                        .build())
                .build();
        assertThat(LaneStateProbe.hasAddresses(e)).isTrue();
    }

    @Test
    void hasAddressesReturnsFalseForNullOrEmpty() {
        assertThat(LaneStateProbe.hasAddresses(null)).isFalse();
        Endpoints e = new EndpointsBuilder().withNewMetadata().withName("svc").endMetadata().build();
        assertThat(LaneStateProbe.hasAddresses(e)).isFalse();
    }
}

File: platform/lib-node/src/__tests__/lane-state-probe.test.ts
import type { V1Endpoints } from "@kubernetes/client-node";
import { Registry } from "prom-client";
import { describe, expect, it } from "vitest";
import { LaneStateProbe } from "../observability/lane-state-probe.js";

describe("LaneStateProbe", () => {
  it("registers gauges for both lanes at zero before any update", () => {
    const registry = new Registry();
    const probe = new LaneStateProbe("services", "payment", registry);

    probe.registerGauges();

    expect(probe.laneValue("stable")).toBe(0);
    expect(probe.laneValue("canary")).toBe(0);
  });

  it("setLaneActive toggles gauge value", () => {
    const registry = new Registry();
    const probe = new LaneStateProbe("services", "payment", registry);
    probe.registerGauges();

    probe.setLaneActive("canary", true);
    expect(probe.laneValue("canary")).toBe(1);

    probe.setLaneActive("canary", false);
    expect(probe.laneValue("canary")).toBe(0);
  });

  it("hasAddresses detects populated/empty Endpoints", () => {
    const populated: V1Endpoints = { subsets: [{ addresses: [{ ip: "10.0.0.1" }] }] };
    const empty: V1Endpoints = { subsets: [] };
    expect(LaneStateProbe.hasAddresses(populated)).toBe(true);
    expect(LaneStateProbe.hasAddresses(empty)).toBe(false);
    expect(LaneStateProbe.hasAddresses(undefined)).toBe(false);
  });

  it("gauge has substrate=http, service, lane labels", async () => {
    const registry = new Registry();
    const probe = new LaneStateProbe("services", "payment", registry);
    probe.registerGauges();
    probe.setLaneActive("stable", true);

    const g = await registry.getSingleMetric("canary_lane_active")?.get();
    expect(g?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: { substrate: "http", service: "payment", lane: "stable" },
      }),
    );
  });
});
